buildSrc/src/main/java/org/elasticsearch/gradle/NoticeTask.java

```diff
@@ -0,0 +1,207 @@
+    @OutputFile
+    private File outputFile = new File(getProject().getBuildDir(), "notices/" + getName() + "/NOTICE.txt");
+    private FileTree sources;
+    /**
+     * Directories to include notices from
+     */
+    private List<File> licensesDirs = new ArrayList<File>();
+
+    public NoticeTask() {
+        setDescription("Create a notice file from dependencies");
+        // Default licenses directory is ${projectDir}/licenses (if it exists)
+        File licensesDir = new File(getProject().getProjectDir(), "licenses");
+        if (licensesDir.exists()) {
+            licensesDirs.add(licensesDir);
+        }
+    }
+
+    /**
+     * Add notices from the specified directory.
+     */
+    public void licensesDir(File licensesDir) {
+        licensesDirs.add(licensesDir);
+    }
+
+    public void source(Object source) {
+        if (sources == null) {
+            sources = getProject().fileTree(source);
+        } else {
+            sources = sources.plus(getProject().fileTree(source));
+        }
+
+    }
+
+    public void source(SourceDirectorySet source) {
+        if (sources == null) {
+            sources = source;
+        } else {
+            sources = sources.plus(source);
+        }
+
+    }
+
+    @TaskAction
+    public void generateNotice() throws IOException {
+        StringBuilder output = new StringBuilder();
+        output.append(readFileToString(inputFile, "UTF-8"));
+        output.append("\n\n");
+        // This is a map rather than a set so that the sort order is the 3rd
+        // party component names, unaffected by the full path to the various files
+        final Map<String, File> seen = new TreeMap<String, File>();
+        FileCollection noticeFiles = getNoticeFiles();
+        if (noticeFiles != null) {
+            for (File file : getNoticeFiles()) {
+                String name = file.getName().replaceFirst("-NOTICE\\.txt$", "");
+                if (seen.containsKey(name)) {
+                    File prevFile = seen.get(name);
+                    String previousFileText = readFileToString(prevFile, "UTF-8");
+                    if (previousFileText.equals(readFileToString(file, "UTF-8")) == false) {
+                        throw new RuntimeException(
+                            "Two different notices exist for dependency '" + name + "': " + prevFile + " and " + file
+                        );
+                    }
+                } else {
+                    seen.put(name, file);
+                }
+            }
+        }
+
+        // Add all LICENSE and NOTICE files in licenses directory
+        for (Map.Entry<String, File> entry : seen.entrySet()) {
+            final String name = entry.getKey();
+            File file = entry.getValue();
+            appendFile(file, name, "NOTICE", output);
+            appendFile(new File(file.getParentFile(), name + "-LICENSE.txt"), name, "LICENSE", output);
+        }
```

Comment:
Could abbreviate this, but obviously not important.

```suggestion
        // Add all LICENSE and NOTICE files in licenses directory
        seen.forEach((name, file) -> {
            appendFile(file, name, "NOTICE", output);
            appendFile(new File(file.getParentFile(), name + "-LICENSE.txt"), name, "LICENSE", output);
        });
```